### PR TITLE
Non-determinism Docs (#2732)

### LINKED ISSUE #2732: Mention that GPU reductions are nondeterministic in docs
# The problem

I am trying out the [MNIST for experts tutorial](https://www.tensorflow.org/versions/r0.7/tutorials/mnist/pros/index.html#deep-mnist-for-experts) and I have inconsistent results on the GPU.
### What do I mean by inconsistent?

With the exactly same network parameters (and randomness removed: read below in the post) every time I run the complete train-then-test process the accuracy is slightly different.
### What have I done to visualize this problem?

For each iteration, I have calculated the differences between the variables (weights, biases) from two _independent but identical_ runs and computed the L1 norm of those differences - <br/>
- [plot](http://i.stack.imgur.com/W5PqZ.png) of L1 norm for the first 1000 iterations in steps of 20.

In a consistent world, these differences should be always zero! 
### How did I remove randomness in the code?
- Removed dropout entirely
- added a graph level seed (`tf.set_random_seed(1234)`). With this the variable initialization is deterministic and also any other randomization in the code. 
- The [MNIST for experts tutorial](https://www.tensorflow.org/versions/r0.7/tutorials/mnist/pros/index.html#deep-mnist-for-experts) uses [this script](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/datasets/mnist.py) to download/load the MNIST data. I have added `numpy.random.seed(3)` in `DataSet.__init__(self, images, labels, fake_data=False, one_hot=False, dtype=dtypes.float32)` in this script to remove randomness during the shuffling process (line 154 in `DataSet.next_batch(self, batch_size, fake_data=False)`)
- `config = tf.ConfigProto(inter_op_parallelism_threads=1, intra_op_parallelism_threads=1)` which goes into the creation of session as `sess = tf.Session(config=config)`
### What system am I using?
- tensorflow 0.8 gpu version (installed via pip)
- OpenSUSE LEAP 42.1 (x86_64)
- Cuda Toolkit 7.5
- CuDNN 4.0
- Tesla K20c card with Nvidia driver 352.79

### PR DESCRIPTION
**Pull request by jkschin**
Fixes https://github.com/tensorflow/tensorflow/issues/2732. Wrote a short tutorial on non-determinism in TensorFlow due to GPU reductions.

### CODE REVIEW

**Comment by michaelisard** [comment]
This looks fine to me, but I will pass to @zheng-xq in case there is a different best practice to ensure determinism when it is required.

**Comment by zheng-xq** [comment]
Erich is working on a design to enable better deterministic support. Adding him to make sure the language is consistent with the direction we are taking.

**Comment by jkschin** [comment]
@ekelsen any updates?

**Comment by ekelsen** [comment]
Reductions will soon be deterministic on the GPU without performance loss.  For some types (float16, complex64, complex128, bool) the performance will increase by 30-1000x.

**Comment by ekelsen** [comment]
This is an awesome doc and thank you for taking the time to write it.  The non-deterministic nature of the GPU reductions has been confusing people for a long time.

However, I'm not sure that it makes sense to pull this in for the brief period of time before the new reductions go in.

**Comment by jkschin** [comment]
Sounds awesome, deterministic reductions with no performance loss. If determinism is guaranteed from then on, then perhaps it doesn't make sense to pull this in for the brief period.

It might make sense to still pull this in and add more explanation on how your side has made it deterministic with no performance loss, for those who intend to understand it.

Feel free to close this PR if it's the best option at this point.

**Comment by drpngx** [comment]
@ekelsen I am closing this PR since it's going to be obsolete pretty soon. Feel free to re-open if you think it's worth having in the meantime.
@tfboyd just FYI
@jkschin nice doc! I wish we had earlier :-)

**Comment by jkschin** [comment]
@drpngx cheers! Shall look for the next PR to do :)

**Comment by alexklibisz** [comment]
@ekelsen Hi, do you maybe have an update for when the deterministic functionality will be available? Is it something implemented in TF itself or a feature of CUDNN? Thank you

**Comment by ekelsen** [comment]
Reductions are deterministic now. l2loss is deterministic.  softmax will be shortly.  cross entropy sometime soon.

**Comment by ekelsen** [comment]
The backward pass of the convolutions is not deterministic, but this basically NVIDIA's fault.  There is possibly to make it deterministic, but it is 3x slower.

**Comment by duncanriach** [comment]
For up-to-date status of work on making TensorFlow operate deterministically on GPUs, please see the following repo: https://github.com/NVIDIA/tensorflow-determinism.
